Add support for NBI fetching AMOs (#247)
openshmem-org/specification: content/shmem_atomic_swap_nbi.tex

\end{C11synopsis}
where \TYPE{} is one of the extended \ac{AMO} types specified by
Table~\ref{extamotypes}.

\begin{Csynopsis}
void @\FuncDecl{shmem\_\FuncParam{TYPENAME}\_atomic\_swap\_nbi}@(TYPE *fetch, TYPE *dest, TYPE value, int pe);
void @\FuncDecl{shmem\_ctx\_\FuncParam{TYPENAME}\_atomic\_swap\_nbi}@(shmem_ctx_t ctx, TYPE *fetch, TYPE *dest, TYPE value, int pe);
\end{Csynopsis}
where \TYPE{} is one of the extended \ac{AMO} types and has a corresponding
\TYPENAME{} specified by Table~\ref{extamotypes}.

\begin{apiarguments}
  \apiargument{IN}{ctx}{A context handle specifying the context on which to
      perform the operation. When this argument is not provided, the
      operation is performed on the default context.}
  \apiargument{OUT}{fetch}{Local data object to be updated.}
  \apiargument{OUT}{dest}{The remotely accessible data object to be
    updated on the remote \ac{PE}.}
  \apiargument{IN}{value}{The value to be atomically written to the remote
    \ac{PE}.}
  \apiargument{IN}{pe}{An integer that indicates the \ac{PE} number on which
    \dest{} is to be updated.}
\end{apiarguments}

\apidescription{

>     The nonblocking \FUNC{shmem\_atomic\_swap\_nbi} routines performs an atomic

Reviewer: ```suggestion
    The nonblocking \FUNC{shmem\_atomic\_swap\_nbi} routines perform an atomic
```

>     swap operation. This routine returns after posting the operation. The

Reviewer 2: I wonder if it would be better to say "initiating" rather than "posting" since posting has a particular meaning in PCIe.

    operation is considered complete after a subsequent call to

>     \FUNC{shmem\_quiet}. At the completion of \FUNC{shmem\_quiet}, it writes

Reviewer: ```suggestion
    \FUNC{shmem\_quiet}. At the completion of \FUNC{shmem\_quiet}, it has written
```

>     \VAR{value} into \dest{} on \ac{PE} and fetches the prior contents of

Reviewer: ```suggestion
    \VAR{value} into \dest{} on \ac{PE} and fetched the prior contents of
```
